stat/distmat/permutation_test.go

```diff
@@ -0,0 +1,62 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package distmat
+
+import (
+	"testing"
+
+	"golang.org/x/exp/rand"
+
+	"gonum.org/v1/gonum/floats"
+	"gonum.org/v1/gonum/mat"
+)
+
+func TestUniformPermutation(t *testing.T) {
+	up := NewUniformPermutation(rand.NewSource(1))
+	for _, n := range []int{10, 32, 64, 100} {
+		m := mat.NewDense(n, n, nil)
+		if m == nil {
+			t.Error("Matrix failed")
+		}
+		up.PermTo(m)
+		if !confirmEachRowAndColumnOneNZ(m) {
+			t.Error("didnt get back a permutation matrix")
+		}
+	}
+
+}
```

Comment:
```
func TestUniformPermutation(t *testing.T) {
	up := NewUniformPermutation(rand.NewSource(1))
	for _, n := range []int{10, 32, 64, 100} {
		m := mat.NewDense(n, n, nil)
		up.PermTo(m)
		for i := 0; i < n; i++ {
			checkPerm(t, "row", i, mat.Row(nil, i, m))
			checkPerm(t, "col", i, mat.Col(nil, i, m))
		}
	}
}

func checkPerm(t *testing.T, dir string, n int, v []float64) {
	var sum float64
	for i, x := range v {
		switch x {
		case 0, 1:
			sum += x
		default:
			t.Errorf("unexpected value in %s %d position %d: %v", dir, n, i, v)
		}
	}
	if sum != 1 {
		t.Errorf("%s %d is not a valid vector: %v", dir, n, v)
	}
}
```

Reply:
This doesn't check that the 1 appears in a different position in every row / column.

Reply:
> this doesn't check that the 1 appears in a different position in every row / column

I don't think that's the case; this tests that every row and every column sums to one which is not possible if there is more than a single unit element in any row or column if only unit and zero elements are allowed which is also checked. Can you show me a counter example that this doesn't find?

Reply:
I mean that the test is correct, but not terribly explicit.

Reply:
With the change of name of the helper I think it meets the usual Gonum test standards.

Reply:
I guess you're right, the comment inside the loop is very clear. LGTM.

Reply:
Could also check that `var ppT mat.Dense; ppT.Mul(up, up.T())` gives you a `ppT` that is equal to I, but at that point we are testing mathematics rather than the implementation.